platform/wayland/cog-utils-wl.c

```diff
@@ -187,6 +187,180 @@ cog_wl_display_destroy(CogWlDisplay *display)
+
+    popup->configured = false;
+
+    popup->option_menu = option_menu;
+
+    popup->width = platform->window.width;
+    popup->height = cog_popup_menu_get_height_for_option_menu(option_menu);
+
+    popup->popup_menu =
+        cog_popup_menu_create(option_menu, display->shm, popup->width, popup->height, display->current_output->scale);
+
+    popup->wl_surface = wl_compositor_create_surface(display->compositor);
+    g_assert(popup->wl_surface);
+
+#ifdef WL_SURFACE_SET_BUFFER_SCALE_SINCE_VERSION
+    if (wl_surface_get_version(popup->wl_surface) >= WL_SURFACE_SET_BUFFER_SCALE_SINCE_VERSION)
+        wl_surface_set_buffer_scale(popup->wl_surface, display->current_output->scale);
+#endif /* WL_SURFACE_SET_BUFFER_SCALE_SINCE_VERSION */
+
+    if (display->xdg_shell != NULL) {
+        popup->xdg_positioner = xdg_wm_base_create_positioner(display->xdg_shell);
+        g_assert(popup->xdg_positioner);
+
+        xdg_positioner_set_size(popup->xdg_positioner, popup->width, popup->height);
+        xdg_positioner_set_anchor_rect(popup->xdg_positioner, 0, (platform->window.height - popup->height),
+                                       popup->width, popup->height);
+
+        popup->xdg_surface = xdg_wm_base_get_xdg_surface(display->xdg_shell, popup->wl_surface);
+        g_assert(popup->xdg_surface);
+
+        static const struct xdg_surface_listener xdg_surface_listener = {.configure = xdg_surface_on_configure};
+        xdg_surface_add_listener(popup->xdg_surface, &xdg_surface_listener, popup);
+        popup->xdg_popup =
+            xdg_surface_get_popup(popup->xdg_surface, platform->window.xdg_surface, popup->xdg_positioner);
+        g_assert(popup->xdg_popup);
+
+        static const struct xdg_popup_listener xdg_popup_listener = {
+            .configure = xdg_popup_on_configure,
+            .popup_done = xdg_popup_on_popup_done,
+        };
+        xdg_popup_grab(popup->xdg_popup, display->seat_default->seat, cog_wl_seat_get_serial(display->seat_default));
+        xdg_popup_add_listener(popup->xdg_popup, &xdg_popup_listener, popup);
+        wl_surface_commit(popup->wl_surface);
+    } else if (display->shell != NULL) {
+        popup->shell_surface = wl_shell_get_shell_surface(display->shell, popup->wl_surface);
+        g_assert(popup->shell_surface);
+
+        wl_shell_surface_add_listener(popup->shell_surface, &shell_popup_surface_listener, NULL);
+        wl_shell_surface_set_popup(popup->shell_surface, display->seat_default->seat,
+                                   cog_wl_seat_get_serial(display->seat_default), platform->window.wl_surface, 0,
+                                   (platform->window.height - popup->height), 0);
+
+        cog_wl_popup_display(popup);
+    }
+
+    return popup;
+}
+
+void
+cog_wl_popup_destroy(CogWlPopup *popup)
+{
+    g_debug("%s: Destroying @ %p", G_STRFUNC, popup);
+    g_assert(popup);
+
+    if (popup->option_menu == NULL)
+        return;
```

Comment:
This check is not needed. All the `g_clear_{pointer,object}` macro uses below already do checks. Also, this check might prevent members from a half-initialized object being freed below and cause memory leaks&mdash;while unlikely, it's possible.

```suggestion
```

I know this is code moved in the refactoring, so we can leave it as-is for now, and do a separate patches later on with cleanups.

Reply:
But still should be needed for the webkit_option_menu_close() function invocation, isn't it?.

Reply:
That's right, good catch.  We need a non-NULL check before calling `webkit_option_menu_close()` but also remove the `return` statement so deallocation continues 👍🏼

Reply:
Fix submitted in a separated PR: https://github.com/Igalia/cog/pull/643